hoomd/md/VolumeConservationMeshForceCompute.cc

```diff
@@ -0,0 +1,374 @@
+// Copyright (c) 2009-2024 The Regents of the University of Michigan.
+// Part of HOOMD-blue, released under the BSD 3-Clause License.
+
+#include "VolumeConservationMeshForceCompute.h"
+
+#include <iostream>
+#include <stdexcept>
+
+using namespace std;
+
+/*! \file VolumeConservationMeshForceCompute.cc
+    \brief Contains code for the VolumeConservationMeshForceCompute class
+*/
+
+namespace hoomd
+    {
+namespace md
+    {
+/*! \param sysdef System to compute forces on
+    \param meshdef Mesh triangulation
+    \param ignore_type boolean whether to ignore types
+    \post Memory is allocated, and forces are zeroed.
+*/
+VolumeConservationMeshForceCompute::VolumeConservationMeshForceCompute(
+    std::shared_ptr<SystemDefinition> sysdef,
+    std::shared_ptr<MeshDefinition> meshdef,
+    bool ignore_type)
+    : ForceCompute(sysdef), m_mesh_data(meshdef), m_ignore_type(ignore_type)
+    {
+    m_exec_conf->msg->notice(5) << "Constructing VolumeConservationMeshForceCompute" << endl;
+
+    unsigned int n_types = m_mesh_data->getMeshTriangleData()->getNTypes();
+
+    if (m_ignore_type)
+        n_types = 1;
+
+    GPUArray<Scalar2> params(n_types, m_exec_conf);
+    m_params.swap(params);
+
+    m_volume = new Scalar[n_types];
```

Comment:
`m_volume` is duplicated as `m_volume_GPU` in the subclass. Use `GPUArray<Scalar> m_volume` in the base class and then there is no need to duplicate code.